File: SDK/GalaxySDK/Development/Samples/C++/GxAutoReconnect/GxAutoReconnect.cpp
#include "stdafx.h"
#include <iostream>
#include "GalaxyIncludes.h"

using namespace std;


//--------------------------------------------------
/**
\brief     用户继承断线处理类
*/
//--------------------------------------------------
class DeviceDisconnectEventHandler : public IDeviceDisconnectEventHandler
{
public:
	virtual ~DeviceDisconnectEventHandler(void) {};

	//--------------------------------------------------
	/**
	\brief     断线回调函数
	\param     pUserParam      用户参数

	\return    void
	*/
	//--------------------------------------------------
	virtual void DoOnDeviceDisconnectEvent(void* pUserParam)
	{
		std::cout << "The disconnect event is triggered!" << std::endl;
	}
};

//--------------------------------------------------
/**
\brief     用户继承重连处理类
*/
//--------------------------------------------------
class DeviceReconnectEventHandler : public IDeviceReconnectEventHandler
{
public:
	virtual ~DeviceReconnectEventHandler(void) {};

	//--------------------------------------------------
	/**
	\brief     重连回调函数
	\param     pUserParam      用户参数

	\return    void
	*/
	//--------------------------------------------------
	virtual void DoOnDeviceReconnectEvent(void* pUserParam)
	{
		std::cout << "The reconnect event is triggered!" << std::endl;
	}
};


int main(int argc, char* argv[])
{
	try
	{
		//初始化设备库
		IGXFactory::GetInstance().Init();

		//枚举相机设备
		GxIAPICPP::gxdeviceinfo_vector vectorDeviceInfo;
		IGXFactory::GetInstance().UpdateDeviceList(1000, vectorDeviceInfo);

		//判断当前设备连接个数
		if (vectorDeviceInfo.size() <= 0)
		{
			cout << "No device!" << endl;
			return 0;
		}

		//通过SN打开相机设备
		CGXDevicePointer pDevice = IGXFactory::GetInstance().OpenDeviceBySN(vectorDeviceInfo[0].GetSN(), GX_ACCESS_EXCLUSIVE);
		//获取相机属性控制对象
		CGXFeatureControlPointer pRemoteControl = pDevice->GetRemoteFeatureControl();
		//流层对象
		CGXStreamPointer pStream;
		if (pDevice->GetStreamCount() > 0)
		{
			pStream = pDevice->OpenStream(0);
		}
		else
		{
			cout << "Not find stream!";
			return 0;
		}

		//选择默认参数组
		pRemoteControl->GetEnumFeature("UserSetSelector")->SetValue("Default");
		//加载参数组
		pRemoteControl->GetCommandFeature("UserSetLoad")->Execute();

		bool bImplemented = pDevice->GetFeatureControl()->IsImplemented("EnableAutoConnection");
		if (!bImplemented)
		{
			printf("该相机不支持断线重连功能!\n");
			//关闭流
			pStream->Close();
			//关闭相机设备
			pDevice->Close();
			//关闭设备库
			IGXFactory::GetInstance().Uninit();
			return 0;
		}

		//开启流层采集
		pStream->StartGrab();
		
		//开启相机采集
		pRemoteControl->GetCommandFeature("AcquisitionStart")->Execute();

		DeviceReconnectEventHandler hReconnectHandle;
		DeviceDisconnectEventHandler hDisconnectHandle;

		//注册重连和掉线回调
		pDevice->RegisterDeviceReconnectCallback(&hReconnectHandle, NULL);
		pDevice->RegisterDeviceDisconnectCallback(&hDisconnectHandle, NULL);


		pDevice->GetFeatureControl()->GetBoolFeature("EnableAutoConnection")->SetValue(true);

		printf("请手动插拔相机触发掉线，测试完成后点击回车完成测试!\n");
		getchar();

		//相机停止采集
		pRemoteControl->GetCommandFeature("AcquisitionStop")->Execute();
		//流层停止采集
		pStream->StopGrab();

		pDevice->UnregisterDeviceDisconnectCallback();
		pDevice->UnregisterDeviceReconnectCallback();

		//关闭流
		pStream->Close();
		//关闭相机设备
		pDevice->Close();
		//关闭设备库
		IGXFactory::GetInstance().Uninit();
	}
	catch (CGalaxyException &e)
	{
		cout << "<" << e.GetErrorCode() << ">" << "<" << e.what() << ">" << endl;
	}
	catch (std::exception &e)
	{
		cout << "<" << e.what() << ">" << endl;
	}

	cout << "<App exit!>" << endl;

	return 0;
}

File: SDK/GalaxySDK/Development/Samples/CSharp/x64/src/GxAutoReconnect/GxAutoReconnect.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading;
using GxIAPINET;

namespace GxAutoConnect
{
    class Sample
    {

        /// <summary>
        /// 重连回调函数
        /// </summary>
        /// <param name="pUserParam">用户私有参数</param>
        private static void __OnDeviceReconnectCallbackFun(object pUserParam)
        {
            Console.Write("The reconnect callback is triggered!\n");
        }

        /// <summary>
        /// 断线回调函数
        /// </summary>
        /// <param name="pUserParam">用户私有参数</param>
        private static void __OnDeviceDisconnectCallbackFun(object pUserParam)
        {
            Console.Write("The disconnect callback is triggered!\n");
        }

        static public IGXFeatureControl m_remoteFeatureControl = null;

        static void Main(string[] args)
        {
            try
            {
                // initialization
                IGXFactory.GetInstance().Init();

                // Enumerating cameras
                List<IGXDeviceInfo> listGxDeviceInfo = new List<IGXDeviceInfo>();
                IGXFactory.GetInstance().UpdateAllDeviceList(300, listGxDeviceInfo);
                if (listGxDeviceInfo.Count < 1)
                {
                    Console.WriteLine("Device not found");
                    IGXFactory.GetInstance().Uninit();
                    Console.WriteLine("Press any key to exit...");
                    Console.ReadKey();
                    return;
                }

                Int32 nDeviceCount = listGxDeviceInfo.Count;
                Console.WriteLine("Found Device: " + nDeviceCount);

                Int32 index = 0;
                foreach (IGXDeviceInfo deviceInfo in listGxDeviceInfo)
                {
                    Console.WriteLine($"\tID: {index}, SN: {deviceInfo.GetSN()}");
                    Console.WriteLine();
                    ++index;
                }

                index = 0;
                string deviceSN = listGxDeviceInfo[index].GetSN();
                IGXDevice objDevice = IGXFactory.GetInstance().OpenDeviceBySN(deviceSN, GX_ACCESS_MODE.GX_ACCESS_CONTROL);
                m_remoteFeatureControl = objDevice.GetRemoteFeatureControl();

                // Restore default parameter group
                m_remoteFeatureControl.GetEnumFeature("UserSetSelector").SetValue("Default");
                m_remoteFeatureControl.GetCommandFeature("UserSetLoad").Execute();

                // Open the specified flow channel
                IGXStream objStream = objDevice.OpenStream(0);

                bool bImplemented = objDevice.GetFeatureControl().IsImplemented("EnableAutoConnection");
                if(!bImplemented)
                {
					Console.WriteLine("该相机不支持断线重连功能!");
                    objStream.Close();
                    objDevice.Close();
                    IGXFactory.GetInstance().Uninit();
                    return;
                }

                objDevice.GetFeatureControl().GetBoolFeature("EnableAutoConnection").SetValue(true);
                DeviceReconnectDelegate reconnectDelegate =
                new DeviceReconnectDelegate(__OnDeviceReconnectCallbackFun);
                DeviceDisconnectDelegate disconnectDelegate =
                new DeviceDisconnectDelegate(__OnDeviceDisconnectCallbackFun);
                
                // Register reconnect callback
                objDevice.RegisterDeviceReconnectCallback(null, reconnectDelegate);
                // Register disconnect callback
                objDevice.RegisterDeviceDisconnectCallback(null, disconnectDelegate);

                Console.WriteLine("请手动插拔相机触发掉线，测试完成后点击回车完成测试!\n");
                Console.ReadKey();

                // Unregister reconnect callback
                objDevice.UnregisterDeviceReconnectCallback();
                // Unregister disconnect callback
                objDevice.UnregisterDeviceDisconnectCallback();

                objStream.Close();
                objDevice.Close();
                IGXFactory.GetInstance().Uninit();
            }
            catch (CGalaxyException ex)
            {
                Console.WriteLine("GalaxyException: " + ex.Message);
            }
            catch (Exception ex)
            {
                Console.WriteLine("Exception: " + ex.Message);
            }

            Console.WriteLine("Press any key to exit...");
            Console.ReadKey();
        }

    }
}